### PR TITLE
Updated to use runOnUiThread for communication between plugin and UI

### PR DESCRIPTION
**Pull request by alzap**

### CODE REVIEW

File: src/android/BackgroundGpsPlugin.java [hunk 1]
```diff
@@ -47,129 +47,196 @@
...
-        return result;
-    }
+	private void init() throws JSONException {
+		if ((activity == null) || (context == null) || (updateServiceIntent == null) || (driveDetectionServiceIntent == null)) {
+			activity = cordova.getActivity();
+			context = activity.getApplicationContext();
+			updateServiceIntent = new Intent(context, TraxLocationUpdateService.class);
+			driveDetectionServiceIntent = new Intent(context, TraxLocationDriveDetectionService.class);
+		}
+	}
```
Comment (steaks) [inline src/android/BackgroundGpsPlugin.java]: Ok so this ensures that we only initialize the activity, updateServiceIntent, and driveDetectionServiceIntent only once.  Was initializing them more than once creating problems?  Or is this just cleanup?

Also, do we need to explicitly check that all 4 are null?  Could we use an isInitialized boolean?

File: src/android/BackgroundGpsPlugin.java [hunk 2]
```diff
@@ -47,129 +47,196 @@
...
+
+	@Override
+	public boolean execute(String action, JSONArray data, final CallbackContext callbackContext) throws JSONException {
+		init();
+		Boolean retVal = false;
+
+		if (ACTION_START.equalsIgnoreCase(action) && !isEnabled) {
+			retVal = true;
+			if (params == null || headers == null || url == null) {
+				activity.runOnUiThread(new Runnable() {
```
Comment (steaks) [inline src/android/BackgroundGpsPlugin.java]: Do we need to run on the UI thread?  It looks like by default cordova tells us that we're running on the "WebCore" thread (not sure exactly what that means...guess it's the "core" thread for the web process).  If we don't want to block the app and be more performant maybe we could use

```
cordova.getThreadPool().execute(new Runnable() {
                public void run() {
                    ...
                    callbackContext.success(); // Thread-safe.
                }
            });
```

http://cordova.apache.org/docs/en/4.0.0/guide_platforms_android_plugin.md.html

Comment (steaks) [inline src/android/BackgroundGpsPlugin.java]: Ok so it looks like the main changes here are that you run everything on the UI thread.  Why is this necessary?  It is necessary because we need to wake the app up?

File: src/android/TraxLocationDriveDetectionService.java [hunk 3]
```diff
@@ -1,42 +1,31 @@
...
+import com.tenforwardconsulting.cordova.bgloc.data.LocationDAO;
+
 public class TraxLocationDriveDetectionService extends Service implements LocationListener {
     private Criteria criteria;
     private LocationManager locationManager;
     private PowerManager.WakeLock wakeLock;
-    private long MIN_TIME_BETWEEN_LOCATION_UPDATES = 30 * 1000; //milliseconds
-    private double MIN_DISTANCE_BETWEEN_LOCATION_UPDATES = 402.336; //meters 402.336 meters ~ 1/4 mile
+	private long MIN_TIME_BETWEEN_LOCATION_UPDATES = 30 * 1000; //milliseconds
+	private double MIN_DISTANCE_BETWEEN_LOCATION_UPDATES = 201.168; //meters 402.336 meters ~ 1/4 mile or 201.168 ~ 1/8 mile
```
Comment (steaks) [inline src/android/TraxLocationDriveDetectionService.java]: Cool, aggressive.  If this is better we should also change IOS.

File: src/android/TraxLocationDriveDetectionService.java [hunk 4]
```diff
@@ -101,11 +104,11 @@ private void setupLocationManager() {
...
     private void setupCriteria() {
         criteria = new Criteria();
-        criteria.setSpeedAccuracy(Criteria.ACCURACY_LOW);
-        criteria.setPowerRequirement(Criteria.POWER_LOW);
-        criteria.setAccuracy(Criteria.NO_REQUIREMENT);
-        criteria.setBearingAccuracy(Criteria.NO_REQUIREMENT);
-        criteria.setVerticalAccuracy(Criteria.NO_REQUIREMENT);
+		criteria.setSpeedAccuracy(Criteria.ACCURACY_LOW);
+		criteria.setPowerRequirement(Criteria.POWER_LOW);
+		criteria.setAccuracy(Criteria.NO_REQUIREMENT);
```
Comment (steaks) [inline src/android/TraxLocationDriveDetectionService.java]: Should we actually require at leas a low accuracy?  Also, I wonder what the implications of setAccuracy vs. setHorizontalAccruacy.  It's confusing, but I think we should use setHorizontalAccuracy(Criteria.ACCURACY_LOW) per the answers in this SO question.

http://stackoverflow.com/questions/17874454/which-is-a-higher-accuracy-criteria-accuracy-high-or-accuracy-fine

File: src/android/TraxLocationDriveDetectionUtil.java [hunk 5]
```diff
@@ -1,25 +1,24 @@
...
-import java.util.Iterator;
+import com.tenforwardconsulting.cordova.bgloc.data.DAOFactory;
+import com.tenforwardconsulting.cordova.bgloc.data.Location;
+import com.tenforwardconsulting.cordova.bgloc.data.LocationDAO;
 
 
 public class TraxLocationDriveDetectionUtil {
     private static String TAG = "TraxLocationDriveDetectionUtil";
-    private final static int SPEEDY_LOCATIONS_THRESHOLD = 2;
+	private final static int SPEEDY_LOCATIONS_THRESHOLD = 1;
```
Comment (steaks) [inline src/android/TraxLocationDriveDetectionUtil.java]: This would mean that one bad reading would trigger the mileage tracker.  I'm not sure we want that.

File: src/android/TraxLocationDriveDetectionUtil.java [hunk 6]
```diff
@@ -1,25 +1,24 @@
...
     private final static int SPEED_LOCATIONS_TIME_WINDOW = 8 * 60 * 1000; //8 minutes
-    private final static double FLOOR = 4.02336; //4.02336 meters/s ~ 9 miles per hour
+	private final static double FLOOR = 2.01168; //4.02336 meters/s ~ 9 MPH or 2.01168 meters/s ~ 4.5 MPH
```
Comment (steaks) [inline src/android/TraxLocationDriveDetectionUtil.java]: pretty aggressive.

**Comment by steaks** [comment]
Cool changes.  Want to talk offline at some point about them?

File: src/android/TraxLocationUpdateService.java [hunk 7]
```diff
@@ -100,10 +86,10 @@ private void setupLocationManager() {
 
     private void setupCriteria() {
         criteria = new Criteria();
-        criteria.setAccuracy(Criteria.ACCURACY_FINE);
-        criteria.setBearingAccuracy(Criteria.NO_REQUIREMENT);
-        criteria.setSpeedAccuracy(Criteria.NO_REQUIREMENT);
-        criteria.setVerticalAccuracy(Criteria.NO_REQUIREMENT);
+		criteria.setAccuracy(Criteria.ACCURACY_HIGH);
```
Comment (steaks) [inline src/android/TraxLocationUpdateService.java]: I think this needs to be setHorizontalAccuracy(Criteria.ACCURACY_HIGH).

http://developer.android.com/reference/android/location/Criteria.html#setAccuracy(int)

File: src/android/BackgroundGpsPlugin.java (hunk 1 above)
Comment (alzap) [inline src/android/BackgroundGpsPlugin.java]: Agreed.

File: src/android/BackgroundGpsPlugin.java (hunk 2 above)
Comment (alzap) [inline src/android/BackgroundGpsPlugin.java]: If we communicate back with the UI, we should use the UI thread.  Otherwise, you will get errors on callbackContext that is expecting an acknowledgment from the UI.  When I run the Debug in Eclipse, all execute() returns an error "cordova plugin attempted to send a second callback for id"

File: src/android/TraxLocationDriveDetectionService.java (hunk 3 above)
Comment (alzap) [inline src/android/TraxLocationDriveDetectionService.java]: Yeah, I will test this more to get a better sense.

File: src/android/TraxLocationDriveDetectionService.java (hunk 4 above)
Comment (alzap) [inline src/android/TraxLocationDriveDetectionService.java]: ok, I'll test it and try it.

File: src/android/TraxLocationDriveDetectionUtil.java (hunk 6 above)
Comment (alzap) [inline src/android/TraxLocationDriveDetectionUtil.java]: Again, this is for testing.  I had a hard time getting it to start.  I guess if you running fast you can accidentally start it.

File: src/android/TraxLocationUpdateService.java (hunk 7 above)
Comment (alzap) [inline src/android/TraxLocationUpdateService.java]: OK, will test it and see if affects it.
